### DISCUSSION TITLE
Using `use` multiple times with `async/await`'s involves?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#23] [Q&A] [answered]
I was wondering if this is the correct way to do it:
```dart
Future<BashTaggedFileUtility> rootBashrcTaggedFileUtility(
  CapsuleHandle use,
) async {
  final fileFuture = use(bashrcFile)(SystemValues.rootUser);
  final contentFuture = use(bashrcContentRootUser);

  return BashTaggedFileUtility(
    file: (await fileFuture)!,
    content: await contentFuture,
  );
}
```

(*) Sorry about naming convention. Maybe I will change it in the future

**Comment by GregoryConrad** [comment, marked as answer]
That looks good!

**Reply by busslina** [reply]
I'm starting to appreciate this lib :)

I read the code for built-in Side Effects and I understand almost everything. I left for the future to read the implementation details. I hope soon I do my first Side Effect with `register`.

BTW, would make any sense to use nested `register/side effects`?

**Reply by GregoryConrad** [reply]
> I'm starting to appreciate this lib :)

glad to hear it! Lmk of any feedback you may have along the way

> I read the code for built-in Side Effects and I understand almost everything.

Well that's good! Wasn't sure if the code behind some of those building-block side effects would make sense to someone that isn't me, especially with how the async oriented side effects are built exclusively with streams (and that is so we can stop listening to them via subscriptions, as for some reason, there is no builtin way to stop listening to a future).

> I hope soon I do my first Side Effect with `register`.

That's really getting into the underlying implementation! I will say that 99% of the time you shouldn't *need* to touch `register`. The idea is to compose side effects together and not implement everything you need from scratch every time (which is what `register` lets you do). That being said, if you just want to experiment, by all means go for it.

Also not sure if I documented it explicitly anywhere, so here goes: `register` simply invokes the supplied callback on a capsule/widget's _first build_ only, and then returns that value from the first build on the first and every subsequent build. The `api` given to the callback is how you actually interface with capsule/widget lifecycles and the underlying container.

> BTW, would make any sense to use nested register/side effects?

Could you please explain a little more what you mean by that? I will say that side effects, at least in this Dart/Flutter implementation, follow the same rules as hooks. I.e., they must be unconditionally called.

So you can do:
```dart
final someFuture = use.memo(() => ..., [...]);
final currState = use.future(someFuture);
```

But you certainly can't do:

```dart
use.effect(() {
  final currState = use.future(...);
}, [...]);
```

If that makes sense.

**Reply by busslina** [reply]
> That's really getting into the underlying implementation! I will say that 99% of the time you shouldn't need to touch register. The idea is to compose side effects together and not implement everything you need from scratch every time (which is what register lets you do). That being said, if you just want to experiment, by all means go for it.

Yeah, I was thinking that. You should have done already almost 99.9% of the basic side effects needed to do anything.

> BTW, would make any sense to use nested register/side effects?

Forget it haha.

The magic behind the scenes still hurts my brain but `1)` I don't need to know that in order to use the lib and `2)` it's matter of time to understand it (for the moment I don't want even to try to read it)

**Reply by busslina** [reply]
Also, I'm surprised  too few people using this lib. Your other lib `Mimir` seems to be more popular (maybe because it's older). I saved it too in my bucket list

**Reply by GregoryConrad** [reply]
> The magic behind the scenes still hurts my brain but 1) I don't need to know that in order to use the lib and 2) it's matter of time to understand it (for the moment I don't want even to try to read it)

If you ever wonder how any of it works, feel free to open another discussion! Happy to lay it all out in more detail. But high level:
- Containers act simply as a `Map<Capsule, CapsuleManager>`
- `CapsuleManager`s houses capsule state, including current data and side effect registrations, and can interact with the container
- `CapsuleHandle` is implemented as a temporary view onto the capsule's `CapsuleManager`
- And then there is a good chunk of fun dependency graph algorithms that you can view here: https://github.com/GregoryConrad/rearch-dart/blob/rearch-v1.3.0/packages/rearch/lib/src/node.dart
- That's about it! Funny enough, the core components only take up a few hundred lines of code each. The idea itself took _forever_ to come to the final design, but once it was designed, the implementation was really streamlined.

> Also, I'm surprised too few people using this lib. Your other lib Mimir seems to be more popular (maybe because it's older). I saved it too in my bucket list

This in part has a few reasons:
1. I am a better developer than I am a marketer
2. There is over-saturation with regards to state management in the front end space. It is hard for a new approach to gain much traction this late in the game, especially with everyone so skeptical of new approaches now.
3. Mimir is the only solution (at least that I am aware of) for a particularly common problem (on-device full-text search suitable for a search bar). I made it when I was frustrated that no library could do what I wanted a year or two back, and I can see that other people are having that problem too. I did think that Mimir would grow faster than it has so far, and again, that is probably related to my not-so-good marketing skills.

In regards to those first two points, I am planning on releasing a blog post (I plan on sharing it on Reddit) in the next few days that should hopefully help bring some exposure to ReArch, but moreso why I made it in the first place. I don't want it to serve just as an advertisement, but rather an analysis onto the current state management landscape, and I think the current draft does that pretty well.

**Reply by busslina** [reply]
Your code is pretty elegant.

> I am a better developer than I am a marketer

At the end the success will belong to you, not to the salesman.

Thanks for the implementation overview.

Keep doing like this :)

**Reply by busslina** [reply]
> Also not sure if I documented it explicitly anywhere, so here goes: register simply invokes the supplied callback on a capsule/widget's first build only, and then returns that value from the first build on the first and every subsequent build. The api given to the callback is how you actually interface with capsule/widget lifecycles and the underlying container.

Seems that the `api` only have one usable method: `rebuild()`. Isn't it or could make sense to use the other two methods by myself?
They are used in `effect()`.

**Reply by GregoryConrad** [reply]
> Seems that the api only have one usable method: rebuild(). Isn't it or could make sense to use the other two methods by myself?

There are 4 methods on the SideEffectApi, see 

https://pub.dev/documentation/rearch/latest/rearch/SideEffectApi-class.html
